Add plugin infra for adding new dashboard tabs/cards

rawagner opened the pull request:
@vojtechszocs @cloudbehl

spadgett commented:
/approve
/assign @vojtechszocs

spadgett commented:
/ok-to-test

rawagner commented:
After discussion with @vojtechszocs I changed this a bit - adding tabs and cards is a separate extension. I also improved types a bit.

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +33,25 @@ namespace ExtensionProperties {
   }
+
+  export interface DashboardsTab {
+    /** The tab's ID which will be used as href within dashboards page */
```
Nit: "as href" => "as part of href".

rawagner replied:
fixed

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +33,25 @@ namespace ExtensionProperties {
+
+  export interface DashboardsCard {
+    /** The tab's ID where this card should be rendered */
+    tabID: string;
```
Can we name this `tab` ? (Similar to `NavItem.perspective` referring to a perspective ID.)

rawagner replied:
sure, renamed to `tab`

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +33,25 @@ namespace ExtensionProperties {
+    position: GridPosition;
+
+    /** The card component to render */
+    component: React.ComponentType<any>;
```
cc @spadgett @christianvogt 

Directly referencing a React component from within a plugin's entry module will make it part of the Console initial chunk, which is already too big.

The default strategy for referencing plugin-contributed React components should be

```ts
Promise<TypeOfComponent>
```

and using either

```ts
() => import('./path/to/module' /* webpackChunkName: "chunk-name" */).then(m => m.MyComponent)
```

or

```ts
async () => (await import('./path/to/module' /* webpackChunkName: "chunk-name" */)).MyComponent
```

to provide those components.

Please add the following into `typings/index.ts`:

```ts
type LazyLoader<T> = () => Promise<React.ComponentType<T>>;
```

and use it in both `typings/pages.ts` and `typings/dashboards.ts`.

rawagner replied:
I cant move `LazyLoader` type to `typings/index.ts` because that results in a circular dependency. Moved it to `typings/types.ts`.

vojtechszocs commented on frontend/public/components/app-contents.tsx:
```diff
@@ -98,7 +98,7 @@ const AppContents = withRouter(React.memo(() => (
           <Route path={['/all-namespaces', '/ns/:ns']} component={RedirectComponent} />
 
-          <LazyRoute path="/dashboards" exact loader={() => import('./dashboards-page/dashboards' /* webpackChunkName: "dashboards" */).then(m => m.DashboardsPage)} />
+          <LazyRoute path="/dashboards" loader={() => import('./dashboards-page/dashboards' /* webpackChunkName: "dashboards" */).then(m => m.DashboardsPage)} />
```
FYI @spadgett plugins can add new tabs into existing Dashboards page, with tab ID reflected into the URL.

vojtechszocs commented on frontend/public/components/dashboards-page/dashboards.tsx:
```diff
@@ -1,20 +1,47 @@
   },
 ];
 
+const getCardsOnPosition = (cards: DashboardsCard[], position: GridPosition): React.ComponentType<any>[] =>
```
To render lazy (async) components, you can use Promise-based `AsyncComponent`.

rawagner replied:
Using `AsyncComponent` now

vojtechszocs commented:
My only concern here is to ensure that we load custom components lazily instead of referencing them directly, please see my inline comments for details.

rawagner commented:
> My only concern here is to ensure that we load custom components lazily instead of referencing them directly, please see my inline comments for details.

Thanks @vojtechszocs for review. I addressed all issues, can you take a look again ?

vojtechszocs commented on frontend/public/components/dashboards-page/dashboards.tsx:
```diff
@@ -2,24 +2,51 @@ import * as React from 'react';
+  return plugins.registry.getDashboardsTabs().map(tab => {
+    const tabCards = cards.filter(c => c.properties.tab === tab.properties.id);
+    return {
+      href: tab.properties.id,
```
We should also write an extension check that ensures tab IDs are unique, I'll put it on my task list.

vojtechszocs commented on frontend/public/components/dashboards-page/dashboards.tsx:
```diff
@@ -2,24 +2,51 @@ import * as React from 'react';
+const getCardsOnPosition = (cards: DashboardsCard[], position: GridPosition): React.ComponentType<any>[] =>
+  cards.filter(c => c.properties.position === position).map(c => () => <AsyncComponent loader={c.properties.loader} />);
+
+const getPluginTabs = (): Page[] => {
```
Nit: I'd rename this to `getPluginTabPages`.

rawagner replied:
ok, renamed

vojtechszocs commented on frontend/public/components/dashboards-page/dashboards.tsx:
```diff
@@ -2,24 +2,51 @@ import * as React from 'react';
       <>
         <PageHeading title="Dashboards" detail={true} />
-        <HorizontalNav match={match} pages={tabs} noStatusBox />
+        <HorizontalNav match={match} pages={tabs.concat(getPluginTabs())} noStatusBox />
```
I'd prefer doing data manipulation outside JSX to keep it simple:

```ts
const tabs: Page[] = [
  {
    href: '',
    name: 'Overview',
    component: OverviewDashboard,
  },
  ...getPluginTabPages(),
];
```

rawagner replied:
done

rawagner commented:
@spadgett can you take a look ?

cloudbehl commented:
@spadgett @rawagner Blocked on this PR to be merged. Can we merge this PR?

jelkosz commented:
@cloudbehl you should not really be blocked, the API should not change so it is safe to base your work on this already

cloudbehl commented:
@jelkosz ok, I was waiting for the PR to be merged. I have raised a dependent PR[1] for now once merged I will rebase it. Thanks

[1]. https://github.com/openshift/console/pull/1803

vojtechszocs commented:
/lgtm

rawagner commented:
/retest
